File: 2025-08/2025-08-24/character_battle.py
def battle(my_army, opposing_army):
    if len(my_army) > len(opposing_army):
        return 'Opponent retreated'

    elif len(my_army) < len(opposing_army):
        return 'We retreated'

    else:
        my_army_score = 0
        opposing_army_score = 0

        for i in range(0, len(my_army)):
            if character_strength(my_army[i]) > character_strength(opposing_army[i]):
                my_army_score += 1
            elif character_strength(my_army[i]) < character_strength(opposing_army[i]):
                opposing_army_score += 1

    return 'We won' if my_army_score > opposing_army_score else 'We lost' if my_army_score < opposing_army_score else 'It was a tie'

def character_strength(character):
    CODE_OF_LOWERCASE_A = ord('a')
    LOWERCASE_A_STRENGTH = 1
    LOWERCASE_DIFF = CODE_OF_LOWERCASE_A - LOWERCASE_A_STRENGTH
    CODE_OF_UPPERCASE_A = ord('A')
    UPPERCASE_A_STRENGTH = 27
    UPPERCASE_DIFF = CODE_OF_UPPERCASE_A - UPPERCASE_A_STRENGTH
    strength = 0

    if character.islower():
        strength = ord(character) - LOWERCASE_DIFF

    elif character.isupper():
        strength = ord(character) - UPPERCASE_DIFF

    elif character.isdigit():
        strength = int(character)

    return strength

print(battle("Hello", "World"))
print(battle("pizza", "salad"))
print(battle("C@T5", "D0G$"))
print(battle("kn!ght", "orc"))
print(battle("PC", "Mac"))
print(battle("Wizards", "Dragons"))
print(battle("Mr. Smith", "Dr. Jones"))

# --- TEST SUITE ---

tests_text = '''
1. battle("Hello", "World") should return "We lost".
2. battle("pizza", "salad") should return "We won".
3. battle("C@T5", "D0G$") should return "We won".
4. battle("kn!ght", "orc") should return "Opponent retreated".
5. battle("PC", "Mac") should return "We retreated".
6. battle("Wizards", "Dragons") should return "It was a tie".
7. battle("Mr. Smith", "Dr. Jones") should return "It was a tie".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 2025-08/2025-08-24/characterBattle.js
function battle(myArmy, opposingArmy) {
  if (myArmy.length > opposingArmy.length) 
    return "Opponent retreated";

  else if (myArmy.length < opposingArmy.length)
    return "We retreated";

  else {
    let myArmyScore = 0;
    let opposingArmyScore = 0;

    for (let i = 0; i < myArmy.length; i ++) {
      if (characterStrength(myArmy[i]) > characterStrength(opposingArmy[i])) 
        myArmyScore ++;

      else if (characterStrength(myArmy[i]) < characterStrength(opposingArmy[i]))
        opposingArmyScore ++;
    }

    return myArmyScore > opposingArmyScore ? "We won" : myArmyScore < opposingArmyScore ? "We lost" : "It was a tie";
  }
}

function characterStrength(character) {
  const codeOfLowerCaseA = 'a'.charCodeAt(0);
  const codeOfLowerCaseZ = 'z'.charCodeAt(0);
  const lowerCaseAStrength = 1;
  const lowerCaseDiff = codeOfLowerCaseA - lowerCaseAStrength;
  const codeOfUpperCaseA = 'A'.charCodeAt(0);
  const codeOfUpperCaseZ = 'Z'.charCodeAt(0);
  const upperCaseAStrength = 27;
  const upperCaseDiff = codeOfUpperCaseA - upperCaseAStrength;
  let strength = 0;

  if (codeOfLowerCaseA <= character.charCodeAt(0) && character.charCodeAt(0) <= codeOfLowerCaseZ)
    strength = character.charCodeAt(0) - lowerCaseDiff;
  
  else if (codeOfUpperCaseA <= character.charCodeAt(0) && character.charCodeAt(0) <= codeOfUpperCaseZ) 
    strength = character.charCodeAt(0) - upperCaseDiff;
  
  else if (0 <= character && character <= 9)
    strength = character;
  return strength;
}

console.log(battle("Hello", "World"));
console.log(battle("pizza", "salad"));
console.log(battle("C@T5", "D0G$"));
console.log(battle("kn!ght", "orc"));
console.log(battle("PC", "Mac"));
console.log(battle("Wizards", "Dragons"));
console.log(battle("Mr. Smith", "Dr. Jones"));
//console.log(battle("Aa2;", "Bb1:"));

// --- TEST SUITE ---

const testsText = `
1. battle("Hello", "World") should return "We lost".
2. battle("pizza", "salad") should return "We won".
3. battle("C@T5", "D0G$") should return "We won".
4. battle("kn!ght", "orc") should return "Opponent retreated".
5. battle("PC", "Mac") should return "We retreated".
6. battle("Wizards", "Dragons") should return "It was a tie".
7. battle("Mr. Smith", "Dr. Jones") should return "It was a tie".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);